projects/cm_mcu/LocalTasks.c

```diff
@@ -708,37 +1202,47 @@ int init_load_clk(int clk_n)
 {
 
   while (getPowerControlState() != POWER_ON) {
-    vTaskDelay(pdMS_TO_TICKS(10)); //delay 10 ms
+    vTaskDelay(pdMS_TO_TICKS(10)); // delay 10 ms
   }
 
   char *clk_ids[5] = {"r0a", "r0b", "r1a", "r1b", "r1c"};
   uint8_t i2c_addrs = CLOCK_CHIP_COMMON_I2C_ADDR; // i2c address of a clock chip
   if (clk_n == 0)
     i2c_addrs = CLOCK_CHIP_R0A_I2C_ADDR;
 
+  // grab the semaphore to ensure unique access to I2C controller
+  if (clock_args.xSem != NULL) {
+    while (xSemaphoreTake(clock_args.xSem, (TickType_t)10) == pdFALSE)
+      ;
+  }
+
```

Comment:
same comment about semaphore here and below.

Reply:
Actually, should I add code to check !NULL for dcdc_args.xSem` too?

Reply:
I think we should just assume that the semaphore exists (get rid of the NULL checks everywhere)

Reply:
ha Okay.

Reply:
should I remove this check even before SemTake? and is there a scenario that a program might be stuck because certain process doesn't let go of semaphore after crashing and rebooting can't resolve it?

Reply:
no, the semaphore is software only. also there is a check for non-null in the SemTake too if you dig into the FreeRTOS code. The way we have the code now (check on the SemTake, don't check on the SemGive) is inconsistent (you get a crash on the release but not on the acquisition?). Previously we didn't use semaphores for all I2C access so since we do now the semaphore should always exist.

Reply:
Okay, got it. (Yes, I think the crash earlier was because of the release alone). Done.